lib/awful/screenshot.lua

```diff
@@ -0,0 +1,884 @@
+
+    directory, prefix = parse_filepath(filepath)
+
+  elseif directory or prefix then
+
+    if directory and type(directory) == "string" then
+      directory = check_directory(directory)
+    elseif self.directory then
+      directory = self._private.directory -- The setter ran check_directory()
+    else
+      directory = get_default_dir()
+    end
+
+    if prefix and type(prefix) == "string" then
+      prefix = check_prefix(prefix)
+    elseif self.prefix then
+      prefix = self._private.prefix -- The setter ran check_prefix()
+    else
+      prefix = get_default_prefix()
+    end
+
+    directory, prefix, filepath = make_filepath(directory, prefix) 
+
+  elseif self.filepath and check_filepath(self.filepath) then
+
+    filepath = self.filepath
+    directory, prefix = parse_filepath(filepath)
+
+  else
+
+    if directory and type(directory) == "string" then
+      directory = check_directory(directory)
+    elseif self.directory then
+      directory = self._private.directory -- The setter ran check_directory()
+    else
+      directory = get_default_dir()
+    end
+
+    if prefix and type(prefix) == "string" then
+      prefix = check_prefix(prefix)
+    elseif self.prefix then
+      prefix = self._private.prefix -- The setter ran check_prefix()
+    else
+      prefix = get_default_prefix()
+    end
+
+    directory, prefix, filepath = make_filepath(directory, prefix) 
+
+  end
+
+  if filepath then
+    self._private.directory = directory -- These have already
+    self._private.prefix = prefix       -- been checked.
+    self._private.filepath = filepath
+  end
+
+end
+
+--- Save screenshot
+--
+-- @function awful.screenshot:save
+-- @tparam[opt] table args Table with the filepath parameters
+function screenshot:save(args)
+
+  self:filepath_builder(args)
+
+  if self._private.surface and self.filepath then
+    self._private.surface:write_to_png(self.filepath)
+  end
+
+end
+
+--- Screenshot constructor - it is possible to call this directly, but it is
+-- recommended to use the helper constructors, such as awful.screenshot.root
+--
+-- Possible args include:
+--   directory [string]: the path to the screenshot directory
+--   prefix [string]: the prefix to prepend to screenshot file names
+--   frame_color: the color of the frame for a snipper tool
+--   method: the method of screenshot to take (i.e. root window, etc).
+--   screen: the screen for a physical screen screenshot
+--   on_success_cb: the callback to run on the screenshot taken with
+--                  a snipper tool
+--   geometry: a geometry object for a snip screenshot
+--
+-- @constructorfct awful.screenshot.new
+-- @tparam[opt] table args Table with the constructor parameters
+local function new(_, args)
+
+  local args = (type(args) == "table" and args) or {}
+  local ss = gears.object({
+    enable_auto_signals = true,
+    enable_properties = true
+  })
+
+  dir, prfx = configure_path(args.directory, args.prefix)
+
+  local fclr = nil
```

Comment:
frame_color

Reply:
The constructor I intentionally did this with everything to avoid name collision. I thought it would be easier to follow. Maybe the constructor local variables should all be named as '_frame_color' ?

Reply:
or mb try using russian 'a' and just name them the same? ;)